egressgateway: provide a very basic Cell

lmb opened the pull request:
These are the minimal changes needed to get the EGW into a hive Cell. The hacky part is that enabling the EGW is gated on the global option.Config.EnableIPv4EgressGateway which is references from a bunch of places. It's not entirely clear to me how to migrate that toggle yet, and it'll probably require a bunch of changes. I've kept the PR simple in the interests of making this easier to review.

bimmlerd commented on pkg/egressgateway/manager.go:
```diff
@@ -45,6 +54,20 @@ const (
+	InstallRoutes bool
+}
+
+var DefaultConfig = Config{
```
Does this have to be exported? Would prefer it to be priv so it's not possible to mutate it from outside

lmb replied:
It's exported because of the ugly "enable egw" hack, I'll try to get rid of it.

bimmlerd commented on pkg/egressgateway/manager.go:
```diff
@@ -32,6 +34,13 @@ var (
 	zeroIPv4 = net.ParseIP("0.0.0.0")
 )
 
+var Cell = cell.Module(
```
Since this is what code navigation will jump you to as an entry point for this file, would be nice to have a comment here explaining what this cell provides

bimmlerd commented on pkg/egressgateway/manager.go:
```diff
@@ -32,6 +34,13 @@ var (
 
+var Cell = cell.Module(
+	"egressgateway",
+	"IPv4 Egress Gateway",
```
this is usually a one line description of what the cell does

bimmlerd commented on pkg/egressgateway/manager.go:
```diff
@@ -81,18 +104,51 @@ type Manager struct {
+	IdentityAllocator identityCache.IdentityAllocator
+
+	// Lifecycle to attach hooks to. Optional.
+	Lifecycle hive.Lifecycle
```
Is this optional for testing's sake? Seems cleaner to me to construct a hive and get a lifecycle from it in the test.

lmb replied:
It's optional so that I don't have to change all testing call sites, is that what you mean? Integrating hive into the tests causes the PR to balloon again, for little direct benefit IMO.

bimmlerd replied:
that's fine, in that case maybe mention that it's optional for testing, and should be supplied in the usual case

lmb replied:
> Integrating hive into the tests causes the PR to balloon again, for little direct benefit IMO.

Turns out i was full of crap, it's fine. Reworked to start a hive.

bimmlerd replied:
I might be missing something, but I see neither a hive nor a non-optional Lifecycle - did you back out of this again?

lmb replied:
Yes, because hive was blowing up on me in CI a couple of times and I couldn't figure out what was wrong. Jussi has since helped me, but overall I'm not convinced hive is a net gain for tests. It moves a bunch of logic to runtime that in tests is much easier to catch at compile time.

bimmlerd replied:
Alright, fair enough.

bimmlerd commented on daemon/cmd/cells.go:
```diff
@@ -91,6 +92,18 @@ var (
+		cell.Config(egressgateway.DefaultConfig),
+		cell.Provide(
+			func(params egressgateway.Params) *egressgateway.Manager {
+				if !option.Config.EnableIPv4EgressGateway {
```
if there's not too much churn maybe pull this into the egress config and out of option.config. But this might be what you alluded to re minimal changes; fine in that case

lmb replied:
yes, this is what i was referring to. I had a chat with gilberto and maybe there is an easy way to work around this, i'll take another look.

lmb replied:
I've replaced a bunch of checks for `Config.Enable..` with `egressGatewayManager != nil`. That works nicely for the k8swatcher, the daemon is less clear cut. I've put the latter in a separate commit.

We're still left with a bunch of hairy global usage unfortunately:
```
pkg/datapath/loader/base.go
343:	if option.Config.Tunnel == option.TunnelDisabled && option.Config.EnableIPv4EgressGateway {

pkg/k8s/synced/crd.go
56:	if option.Config.EnableIPv4EgressGateway {

pkg/k8s/watchers/cilium_endpoint.go
212:	if option.Config.EnableIPv4EgressGateway {
239:	if option.Config.EnableIPv4EgressGateway {

pkg/datapath/linux/config/config.go
270:	if option.Config.EnableIPv4EgressGateway {
```

I think to fix this in a nicer way we need a way to define flags in hive cells, but somehow plumb them into `option.Config` during the transition phase. Otherwise we need to do gigantic refactors again. cc @joamaki

joamaki replied:
Preferably I'd of course see these in `EgressGatewayConfig`, but seeing where the enable flag is referenced it'd get painful to do that in this PR since you'd need to plumb the EgressGatewayConfig struct to all these places that have not yet been modularized. Let's keep that out from this PR for the time being. 

As a transitionary thing we could already add EgressGatewayConfig, keep the option.Config.EnableIPv4EgressGateway, remove the flag registration from `daemon/cmd/daemon_main.go` and do it in EgressGatewayConfig.Flags() instead, and finally set `option.Config.EnableIPv4EgressGateway` somewhere... and the last point is where it gets a bit hairy. It can be made to work by having a `cell.Invoke` at the very top in `daemon/cmd/cells.go` which would depend on `*DaemonConfig` and `EgressGatewayConfig` and bridge the two before the "daemonCell" or anyone else gets access to it. I don't like it that much though. I think it'd be preferable to just do it cleanly once when we're far enough that we can do it.

Another thing here is that most likely `EgressGatewayConfig` is a "datapath" configuration, e.g. it affects the operation of datapath (above e.g. pkg/datapath/loader/base.go) more than control-plane and it should likely live under `pkg/datapath` and possibly stay internal to it. Control-plane components that manage the "egress gateway datapath" would then depend on the config either directly, or via some "egress gateway datapath API" to tell them what can and cannot be done. This is maybe less clear with egress gateway, but with datapath components that require probing to figure out the "actual" configuration, it makes sense to not expose the user provided config struct, but rather the configuration that's the result of merging user configuration and probe results. 

tl;dr: let's do nothing in this PR about the config :grin:

bimmlerd reviewed (commented):
couple of questions and some docs nit

jibi commented on daemon/cmd/daemon.go:
```diff
@@ -683,8 +685,9 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup,
-		d.egressGatewayManager = egressgateway.NewEgressGatewayManager(cacheStatus, d.identityAllocator, option.Config.InstallEgressGatewayRoutes)
+	var egressGatewayWatcher watchers.EgressGatewayManager
+	if d.egressGatewayManager != nil {
+		egressGatewayWatcher = d.egressGatewayManager
```
Can't we just pass `d.egressGatewayManager` to `watchers.NewK8sWatcher()`?

lmb replied:
Unfortunately not, because of the dreaded `nil` interface bug. Passing directly will construct the following:

```go
var v watchers.EgressGatewayManager = (*Manager)(nil)
// we need
var v watchers.EgressGatewayManager = nil
```

Want me to comment that?

jibi replied:
Makes sense, no need to add a comment

lmb committed:
egressgateway: provide a hive.Cell

Introduce a constructor which takes a hive-compatible Params struct
and wire things up into a global Cell.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>
